### PR TITLE
Improved protocol

### LINKED ISSUE #5: Sample actor supporting atomic execution.
Leverage the [basic primitives](https://github.com/consensus-shipyard/builtin-actors/blob/feat/atomic-exec/primitives/src/atomic/mod.rs) to write actors that support IPC atomic executions to write a sample actor that illustrates how to use these primitives to perform an atomic exec. We can get inspiration from the sample actor we wrote targeting the `LegacyVM`: https://github.com/filecoin-project/eudico/blob/eudico/chain/consensus/actors/atomic-replace/replace.go.

This can be used to test that the implementation of the atomic execution protocol actually works. Over this contract we should write an integration test once the new IPC actors are integrated in Lotus.

### PR DESCRIPTION
**Pull request by sergefdrv**
This pull request attempts to implement an improved version of the atomic execution protocol described [here](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3882477). It aims at replacing https://github.com/consensus-shipyard/builtin-actors/pull/17.

Resolves #5
Closes consensus-shipyard/builtin-actors/pull/17

### CODE REVIEW

**Comment by sergefdrv** [comment]
This PR is based on the [`atomic_execution`](https://github.com/consensus-shipyard/ipc-atomic-execution/tree/atomic_execution) branch and so far only implements the coordinator actor logic. To be extended with unit tests and primitives useful for implementing execution actors.

File: src/state.rs
Comment (adlrocha) [inline src/state.rs]: Why `from_bytes` here? `ConstructParams` is already passing an address, right?

File: src/lib.rs
Comment (adlrocha) [inline src/lib.rs:75]: How do the actors in the different subnets agree on the `exec_id` of the execution?

File: src/types.rs
Comment (adlrocha) [inline src/types.rs:25]: So the `PreCommit` only specifies the `MethodNum` called for the execution, but how can you know the state that needs to be locked? I think you need to provide some additional information like the params for the method at least.

File: src/atomic.rs
Comment (adlrocha) [inline src/atomic.rs]: I don't think these functions match what the coordinator expects, right? I am having a hard time to follow the end-to-end of an execution and the interactions between the atomic-exec actors and the coordinator.

File: src/state.rs
Comment (adlrocha) [inline src/state.rs]: We may be able to simplify this a bit. Having a `Hamt<Hamt>` means that you have to load twice from the blockstore (and there is an overhead to pass the WASM boundary through the syscall). I don't think we'll have a large number of actors involved in an execution, we can probably use a plain old slice or HashMap for entries.

File: src/lib.rs
Comment (adlrocha) [inline src/lib.rs:120]: Now that I see this call (and definitely future work), it may be a good idea to add to the gateway a `batchSend` method, WDYT? Let's add an issue if you feel this could be useful.

File: src/lib.rs
Comment (adlrocha) [inline src/lib.rs:171]: Here we check that we receive a pre-commitment from all actors, but how do we know that they reached the same output? And how to actors know what state to commit?

File: src/lib.rs
Comment (adlrocha) [inline src/lib.rs:63]: We expect the `ipc_address` to be the caller as these messages are expected to be received as cross-net messages, right? (can we add a slight comment here so it is clear? Thanks).

File: src/lib.rs
Comment (adlrocha) [inline src/lib.rs:102]: Shouldn't this be a specific entrypoint to dispatch atomic exec commitments (I don't know if calling the method directly would work)?

**Review by adlrocha — COMMENTED** [review]
Thank you for the PR. @sergefdrv, this is a first pass (happy to review more thoroughly if needed, I see is still a draft so that is why I didn't went deeper).

My main comment is that it is quite hard to follow the end-to-end and the interactions between the actors involved in each subnets and the coordinator. We are either missing things, or I didn't understand the logic fully. Adding a set of unit test may help here.

**Comment by sergefdrv** [comment]
> ... it is quite hard to follow the end-to-end and the interactions between the actors involved in each subnets and the coordinator. We are either missing things, or ...

Right, I only implemented the coordinator actor logic, the rest is to come. At this point I would like to receive some feedback about the general code structure, especially the IPLD part, which I'm not quite sure I'm doing right.

File: src/state.rs
Comment (sergefdrv) [reply src/state.rs]: Hmm, it seems I should adjust the `ConstructParams` struct, which is still defined as

```rust
pub struct ConstructorParams {
    pub ipc_gateway_address: Vec<u8>,
}
```

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs:75]: As described [here](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3882477), it is to be deterministically computed by each of the execution actors such that it is unique for each distinct set of inputs (including actor IDs) and such that all the execution actors compute the same ID.

Comment (sergefdrv) [reply src/lib.rs:75]: Note that in correct execution, each execution actor is supposed to be invoked with the same set of inputs and the same set of execution actor IDs (absolute IPC addresses). If some user messes it up and invokes its execution actor with a wrong set of parameter then the atomic execution will never commit, and will have to be aborted.

File: src/types.rs
Comment (sergefdrv) [reply src/types.rs:25]: The method number here is the method of the execution actor which ultimately commits the atomic execution. The rest is managed by the execution actors themselves: they derive the unique atomic execution ID (`exec_id`) from the parameters of the atomic execution, such as inputs/locks. Note that the coordinator actor handled `exec_id` as a purely opaque identifier - it doesn't attempt to pull the state identified by the CID.

File: src/atomic.rs
Comment (sergefdrv) [reply src/atomic.rs]: Right, I'm going to tackle this next.

File: src/state.rs
Comment (sergefdrv) [reply src/state.rs]: Thanks for pointing this out! :+1:

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs:120]: We could make `SendCross` method always receive a set of messages, but simply supply a singleton set most of the time. WDYT?

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs:171]: The atomic execution ID `exec_id` must uniquely determine the output:

> ... The execution actor further computes the atomic execution output and
stores it in its state associated with the atomic execution ID.

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs:63]: This is mentioned in the doc comment of the method:

> This method is to be invoked by a wrapped crossnet message

But I agree, it's better to add a bit of clarification in the code, as well.

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs:102]: Not sure I got your question exactly. Do you mean reserving fixed method numbers in execution actors to commit/rollback the atomic execution? 

So far I wanted to give the execution actors flexibility and let them specify the callback method number themselves, but we may revise this once [FIP-0042](https://github.com/filecoin-project/FIPs/blob/master/FRCs/frc-0042.md) is finalized.

File: src/state.rs
Comment (sergefdrv) [reply src/state.rs]: Would it also make difference if we use a hash of `RegistryKey` as the key in the registry HAMT?

**Comment by sergefdrv** [comment]
Summary of [changes](https://github.com/consensus-shipyard/ipc-atomic-execution/compare/dd3e1e14a09848d9ca7b094622e215bcb9d98188..4aac275bd53f3c1e97b5e45c4e8e6c311cedae07): using `Address` type in `ConstructorParams`

File: src/state.rs
Comment (sergefdrv) [reply src/state.rs]: I tried to simplify the registry type in 0726536.

File: src/lib.rs
Comment (adlrocha) [reply src/lib.rs:102]: Can you remind me what are the stages of the execution? There at least two methods for each contract involved in the protocol: one for the messages involved in the atomic execution, and one to commit the final state. How I imagine the end-to-end is that:
- A set of actors decide to perform an atomic execution that is represented through a set of messages sent to the actors in different subnets.
- Once this execution is performed and the coordinator actors decides it has been done successfully, the corresponding outputs need to be committed in the state of the actors in each subnet, but what is the method responsible for accepting this commitment message and merging the state, the same as the ones called in the atomic exec?

File: src/lib.rs
Comment (adlrocha) [reply src/lib.rs:63]: Thanks!

File: src/lib.rs
Comment (adlrocha) [reply src/lib.rs:171]: Right, but there is more than one actor involved in the execution, and each of them live in a different subnet, how do they reach the same `exec_id`? I guess that through the input parameters for the execution, right? But this is not obvious as this is not shared in any way with the coordinator (let me know if my point is not clear, I can try to follow-up or we can have a sync)

File: src/lib.rs
Comment (adlrocha) [reply src/lib.rs:120]: I like it. Do you mind tracking it in an issue in the IPC-gateway repo?

File: src/state.rs
Comment (adlrocha) [reply src/state.rs]: > Would it also make difference if we use a hash of RegistryKey as the key in the registry HAMT?

The HAMT handle this for you under the hood. I think the current approach only does a single call to the blockstore and is simpler. Thanks.

File: src/atomic.rs
Comment (adlrocha) [reply src/atomic.rs]: Gotcha. Resolving this in the meantime then.

File: src/types.rs
Comment (adlrocha) [reply src/types.rs:25]: Gotcha, so the underlaying arguments, messages and methods for the execution are implicit in `exec_id` and the implementation of the actor in each subnet needs to deterministically infer this `exec_id`, right? This links with a comment I left below, if the answer to this is "yes"  we can resolve the one below :)

File: src/lib.rs
Comment (adlrocha) [reply src/lib.rs:75]: That is a fair assumption. That's fine.

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs:171]: Exactly, all of the execution actors are supposed to receive the same set of inputs when invoked for pre-commitment:

> The method deterministically computes a concise _atomic execution ID_ which must be unique for each distinct set of inputs (including actor IDs) and such that all the execution actors compute the same ID.

The key point is that the actors must deterministically produce the same `exec_id`. It is _the responsibility of the execution actors and the users_ to ensure that in order to complete the atomic execution successfully.

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs:102]: The happy path roughly looks like follows (more details [here](https://github.com/protocol/ConsensusLab/discussions/154#discussioncomment-3882477)).

- First, each of the execution actors locally initializes a new instance of the atomic execution protocol by locking the relevant potion of its state and producing a unique atomic execution input (`exec_actor.init_atomic_exec(...)`). 
- Next, the users exchange and collect the full set of inputs, which they use to invoke their execution actors for pre-commitment (`exec_actor.prep_atomic_exec(inputs, ...)`). At this stage, each of the execution actors computes the atomic execution ID, produces and tentatively stores some output, then emits a cross-net message to the coordinator actor (`coord_actor.pre_commit(actors, exec_id, commit)`).
- Once the coordinator actor has recorded a pre-commitment from each of the execution actors, it emits a cross-net message for each of the execution actor to commit the atomic execution (`exec_actor.commit(exec_id)`).
- Finally, each of the execution actor unlocks its state and merges the output once received the commit message from the coordinator actor.

File: src/state.rs
Comment (sergefdrv) [reply src/state.rs]: @adlrocha does it look fine now?

Comment (sergefdrv) [reply src/state.rs]: I think it would make sense to use a hash/CID as a key instead of plain pair (`RegistryKey`), because HAMT seems to store the plain key, although we don't seem to really need it.

File: src/atomic.rs
Comment (adlrocha) [inline src/atomic.rs]: What is the purpose of using a `Cell` here? The `lock` should be enough for the mutability pattern you are trying to use here, right? Removing the `Cell` and using `Option<T>` should work in the same way, right?

File: src/atomic.rs
Comment (adlrocha) [inline src/atomic.rs]: OK! So what you are saying here is that you may receive as an input state that this not yet locked and you lock it before computing its locked CID and registering it in the actor, right? And this propagates to the underlying state provided as an input.

What if the state is already locked and used for some other computation?

File: src/atomic.rs
Comment (adlrocha) [inline src/atomic.rs]: `InputState` represents the whole state for the actor or a part of the whole state of the actor that can be locked individually for the atomic execution? (my question will probably be answered in your next PR :) but just in case)

File: src/atomic.rs
Comment (adlrocha) [inline src/atomic.rs]: How would you retrieve the associated data if it leaves in some other subnet?

File: src/atomic.rs
Comment (adlrocha) [inline src/atomic.rs]: We need a better description of what is the role of `input_fn` and `output_fn`, it may make more sense to me once you have the actor example, but so far I don't see how the `output` can be computed here or the role of `input_fn`.

I guess `output_fn` is the atomic execution we want to perform, right? But these should probably be described as a list of messages to execute.

File: src/atomic.rs
Comment (adlrocha) [inline src/atomic.rs]: Same as above. We need better descriptions of what these functions are. To double-check my understanding: 
- Output function computes the output for the messages received as input in the atomic execution. 
- Apply functions takes the output and commits it into the state.

**Review by adlrocha — COMMENTED** [review]
I like where this is going :) There are a few things I still don't fully understand, but feel free to answer my comments now or defer me to your next PR if they are non-sense. Whatever works best for you.

**Comment by sergefdrv** [comment]
Summary of [changes](https://github.com/consensus-shipyard/ipc-atomic-execution/compare/eadbecad89b8c7167b99065267c9e01c283dd9bc..10528ca211d33cf58299430fd18898cf49334298): slightly adjusted code and added comments

File: src/atomic.rs
Comment (sergefdrv) [reply src/atomic.rs]: I'm not quite sure it's worth it, but this cell acts like a transparent cache to avoid recomputing the CID. I used the interior pattern here because the `cid` method receives a shared reference to `self`. (It would be odd to require an exclusive reference to `self` for a conceptually read-only method.)

File: src/atomic.rs
Comment (sergefdrv) [reply src/atomic.rs]: Here I either attempt to lock the state, if asked to do so, or capture its CID, if it is not yet locked. In the latter case, allowing the actor writer to optimistically delay the locking, At the next step, `prepare_atomic_exec` will check the CID before finally locking any state that was not locked before calling `init_atomic_exec`. In any case, attempting to lock an already locked piece of state returns an error.

Comment (sergefdrv) [reply src/atomic.rs]: Ugh, I think there is a bug; it should be:
```rust
            if lock {
                s.lock()?;
            } if !s.is_locked() {
                v.push(s.cid());
```

File: src/atomic.rs
Comment (sergefdrv) [reply src/atomic.rs]: It is assumed to be a piece of the actor state.

File: src/atomic.rs
Comment (sergefdrv) [reply src/atomic.rs]: It is only about the local data which has been previously supplied to `init_atomic_exec`.

File: src/atomic.rs
Comment (sergefdrv) [reply src/atomic.rs]: I added more comments which may help to better understand the role of those closures.

The output is computed from the optional input data associated with actor's own input ID and any values captured by the closure from its environment. The input data may include the messages to execute, whereas the closure's environment would include any additional data from other executing actors that is required to produce the output.

File: src/atomic.rs
Comment (sergefdrv) [reply src/atomic.rs]: This `output_fn` only interprets the serialized (raw) representation of the output produced by the other closure supplied to `prepare_atomic_exec`. Check the added comments which may help to see this better.

File: src/lib.rs
Comment (sergefdrv) [inline src/lib.rs]: I wonder if we should make `atomic` module `pub`, also renaming it to `primitives`? Or even make it a separate create?

Comment (adlrocha) [reply src/lib.rs]: Yep. What about maybe adding an `sdk` crate within this repo where we expose all the basic primitives to be used by actors supporting atomic_exec? In the future we can take this sdk and make it part of the `ipc_sdk` (@cryptoAtwill was going to kick-off the work on this soon).

File: src/atomic.rs
Comment (adlrocha) [reply src/atomic.rs]: Ok, it makes more sense now. I'll wait for unit tests or an example actor to be ready to the final pass on this matter :) Thanks!

File: src/lib.rs
Comment (sergefdrv) [reply src/lib.rs]: I was trying to understand better how Rust packaging system works and stumbled upon this:

> A package can contain as many binary crates as you like, but at most only one library crate. A package must contain at least one crate, whether that’s a library or binary crate.

It seems that we can't make this a separate library crate in this repository unless we convert the repo into a workspace. Have you meant converting this repo into a workspace?

Comment (sergefdrv) [reply src/lib.rs]: Or do you mean following the [root package](https://doc.rust-lang.org/cargo/reference/workspaces.html#root-package) approach?

Comment (adlrocha) [reply src/lib.rs]: No strong preference, I leave it up to you. I was thinking to either add workspace, or even have it as a single library exposing several independent `mod` for namespacing.

Comment (adlrocha) [reply src/lib.rs]: For reference, this is how we do it in `fvm_utils`: https://github.com/consensus-shipyard/fvm-utils/blob/main/Cargo.toml

**Comment by sergefdrv** [comment]
Summary of [changes](https://github.com/consensus-shipyard/ipc-atomic-execution/pull/6#commits-pushed-1c3943b):
 - Adjust the type of exec ID in coordinator actor
 - Using hash as atomic registry key in coordinator actor
 - Derive commonly useful trait impl for `atomic::AtomicInputState`
 - Add `get` and `set` method for `atomic::AtomicInputState`
 - Redesign interface of `atomic::AtomicExecRegistry`

**Comment by sergefdrv** [comment]
I added a sample implementation of a simplistic [FRC-46](https://github.com/filecoin-project/FIPs/blob/master/FRCs/frc-0046.md)-like actor supporting atomic execution. The sample illustrates how the atomic execution primitives can be used.

File: examples/fungible-token/TODO
Comment (adlrocha) [inline examples/fungible-token/TODO]: It would be great if we can include also the build script. You can copy-paste from any of our other actors (actually we should maybe also think about making this part of our sdk).

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [inline examples/fungible-token/src/lib.rs:85]: Why do we pass the `balances` in the constructor params? Wouldn't it be better to initialize this in the constructor?

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [inline examples/fungible-token/src/lib.rs]: typoe?
```suggestion
                let res = Self::prepare_atomic_transfer(rt, cbor::deserialize_params(params)?)?;
```

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [inline examples/fungible-token/src/lib.rs]: typo?
```suggestion
    fn prepare_atomic_transfer<BS, RT>(
```

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [inline examples/fungible-token/src/lib.rs:292]: Cancel can be triggered by a user after initiating and before preparing. After prepare, the only way to cancel an execution is by calling the coordinator to trigger a rollback, right?

File: examples/fungible-token/src/state.rs
Comment (adlrocha) [inline examples/fungible-token/src/state.rs:46]: I see! Are providing a balance table in the constructor so you can set a token pre-allocation to avoid having to implement a minting function and a circulating supply? That makes sense.

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [inline examples/fungible-token/src/lib.rs:245]: Where do you check if the balance for the address is being used for an atomic execution. If I got the code correctly you can still use the balance locked for an execution for an internal transfer (not involved in an atomic exec), or did I missed something?

**Review by adlrocha — CHANGES_REQUESTED** [review]
Looking good! Left a few more comments, let me know if you need a more thorough review somewhere.

It would be great if you could add some comments in the fungible token example (at least in the function headers) to ease the understanding. We want this to be the sample actor that developers use to understand the atomic exec primitives so they can include them in their actors. Also, if we could have some basic unit test coverage it would really help the next one picking up the code. 

The PR is getting a bit large and hard to review by now. What about adding those comments and unit tests and merging this (if we see there is something outstanding for the implementation we can target it in a follow-up PR).

Thanks :pray:

**Comment by adlrocha** [comment]
BTW, for some reason the CI is not happy :shrug: just so you are aware

File: examples/fungible-token/src/lib.rs
Comment (sergefdrv) [reply examples/fungible-token/src/lib.rs:292]: Almost right, one can abort an already prepared atomic execution by invoking the `AbortAtomicTransfer` method of this actor, not the coordinator directly.

File: examples/fungible-token/src/state.rs
Comment (sergefdrv) [reply examples/fungible-token/src/state.rs:46]: Exactly. We could also supply credentials of the owner, who is allowed to mint tokens, but that would be more complicated than necessary for this example.

File: examples/fungible-token/src/lib.rs
Comment (sergefdrv) [reply examples/fungible-token/src/lib.rs:245]: [This](https://github.com/consensus-shipyard/ipc-atomic-execution/pull/6/files/e5d70adfecfd7883382fdb84ece2308027c5617f#diff-9496377190083568b9e1883497f39c8481a75733b86a346f1bbe8d08744cb824R270-R273) will fail if sender's balance is locked for an atomic transfer because `AtomicInputState` doesn't allow modifying the inner state while locked.

File: examples/fungible-token/src/lib.rs
Comment (sergefdrv) [reply examples/fungible-token/src/lib.rs:85]: I believe [this](https://github.com/consensus-shipyard/ipc-atomic-execution/pull/6#discussion_r1061547726) answers your question.

Comment (adlrocha) [reply examples/fungible-token/src/lib.rs:85]: It does, thanks!

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [reply examples/fungible-token/src/lib.rs]: Feel free to resolve when fixed. Thanks!

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [reply examples/fungible-token/src/lib.rs:292]: But the abortion would have to propagated to the coordinator to inform it about this, right? (again, the protocol has a lot of moving parts, it would be great to have some comments to support the code :pray: )

File: examples/fungible-token/src/state.rs
Comment (adlrocha) [reply examples/fungible-token/src/state.rs:46]: This approach works, thank you for the clarification.

File: examples/fungible-token/src/lib.rs
Comment (adlrocha) [reply examples/fungible-token/src/lib.rs:245]: Oh! So the check is implicit in `modify` (I missed that). Got it.

File: examples/fungible-token/src/lib.rs
Comment (sergefdrv) [reply examples/fungible-token/src/lib.rs:292]: There are only two callable methods in the coordinator actor: `PreCommit` and `Revoke`. When cancelling an atomic execution that has not yet been pre-committed by the executing actor, as opposed to aborting an already pre-committed one, there is nothing to revoke at the coordinator actor, so the executing actor just consumes its input ID and releases the lock. After that, the input ID will not be accepted by the actor for pre-commitment anymore.

**Review by adlrocha — APPROVED** [review]
LGTM. Let's merge and migrate to `ipc-actors` so we can start testing them.

**Pull request merged**
